File: matlab_code/Concavity.m
load Af_distance.txt
load Af_elevation.txt
A = Af_distance;
B = Af_elevation;
[m,n] = size(A);
offset_norm_elev = nan(m,n);
median_norm_offset = nan(n,1);
iqr_norm_offset = nan(n,1);
for i = 1:n
    X = A(:,i);
    E = B(:,i);
    R = max(B(:,i))-min(B(:,i));
    x1 = A(1,i);
    x2 = max(A(:,i));
    y1 = B(1,i);
    y2 = max(B(:,i));%where x1,y1 x2,y2 are the long profile endpoints
    x = [x1 x2];
    y = [y1 y2];
    [r m b] = regression(x,y,'one');
    Y = m*X+b; %Y values on the line
    offset_elev = (E-Y)/R;
    offset_norm_elev(:,i) = offset_elev;
    mean_norm_offset(i) = nanmean(offset_elev);
    median_norm_offset(i) = nanmedian(offset_elev);
    iqr_norm_offset(i) = iqr(offset_elev);
end
boxplot(offset_norm_elev,'Notch','on')
set(gca,'YMinorTick','on')
xlabel('River');
ylabel('Normalised concavity index')
axis([0.5,7.5,-1,0.5])
mean(median_norm_offset)
std(median_norm_offset)
stderror=std(median_norm_offset)/sqrt(length(median_norm_offset))

File: matlab_code/concavity.py
import sys
from scipy import stats
import numpy as np
import matplotlib.pyplot as plt

offset_elevs = []
median_norm_offsets = []

output_filename = sys.argv[1]

n = len(sys.argv[2:])

for filename in sys.argv[1:]:

    data = np.genfromtxt(filename, delimiter=',')

    A = data[:, 5]
    B = data[:, 4]

    R = np.nanmax(B) - np.nanmin(B)

    # where x1,y1 x2,y2 are the long profile endpoints
    x = [A[0], np.nanmax(A)]
    y = [B[0], np.nanmax(B)]

    result = stats.linregress(x, y)

    m = result[0]
    b = result[1]

    Y = m * A + b  # Y values on the line
    offset_elev = (B - Y) / R
    offset_elevs.append(offset_elev)

    median_norm_offsets.append(np.nanmedian(offset_elev))

mean = np.nanmean(median_norm_offsets)
std = np.nanstd(np.array(median_norm_offsets))
stderror = std / np.sqrt(n)

print(round(mean, 4), round(std, 4), round(stderror, 4))

plt.boxplot(offset_elevs, notch=True)
plt.xlabel('River')
plt.ylabel('Normalised concavity index')
plt.xlim(0.5, n + 0.5)
plt.ylim(-1, 0.5)
plt.savefig('{}.png'.format(output_filename))
